### PR TITLE
add support for generic methods where the generic types may not be in…

### PR DESCRIPTION
**Pull request by sage-nathan-jones**
… the method's argument list. Maintains backward compatibility with jobs already serialized before this change.

Addresses issue #893

### CODE REVIEW

File: src/Hangfire.Core/Storage/InvocationData.cs [hunk 1]
```diff
@@ -30,17 +30,19 @@ namespace Hangfire.Storage
...
             Method = method;
             ParameterTypes = parameterTypes;
             Arguments = arguments;
+            GenericTypes = genericTypes;
         }
 
         public string Type { get; }
         public string Method { get; }
         public string ParameterTypes { get; }
+        public string GenericTypes { get; }
```
Comment (aidmsu) [inline src/Hangfire.Core/Storage/InvocationData.cs]: Consider not to create separate property `GenericTypes`. Can we store info about generic arguments in `Method` property? `Method` value may look like this:
```
Generic[\"System.Int32\"]
```

File: src/Hangfire.Core/Common/TypeExtensions.cs [hunk 2]
```diff
@@ -70,12 +71,20 @@ public static string ToGenericTypeString(this Type type)
                     continue;
                 }
 
-                var parameterTypesMatched = true;
 
                 var genericArguments = methodCandidate.ContainsGenericParameters
-                    ? new Type[methodCandidate.GetGenericArguments().Length]
-                    : null;
-                
+                        ? new Type[methodCandidate.GetGenericArguments().Length]
```
Comment (aidmsu) [inline src/Hangfire.Core/Common/TypeExtensions.cs]: Could we rename `genericTypes` method parameter to `genericArguments`? It will be able to simplify code. The onle change will be in `genericArguments` assignment:
```
if (methodCandidate.ContainsGenericParameters)
{
    var genericArgumentsCount = methodCandidate.GetGenericArguments().Length;
    if (genericArguments == null) genericArguments = new Type[genericArgumentsCount];
    if (genericArgumentsCount != genericArguments.Length) continue;
}

File: src/Hangfire.Core/Storage/InvocationData.cs (hunk 1 above)
Comment (sage-nathan-jones) [inline src/Hangfire.Core/Storage/InvocationData.cs]: I suppose that it could be done as you suggest, but if I understand the code, then the serialization and deserialization of the InvocationData would become more convoluted. Instead of just parsing a string (method) and JSON array (types) from separate strings in the JSON object, we'd have first take apart the .Method and then deserialize the components separately.

File: src/Hangfire.Core/Common/TypeExtensions.cs (hunk 2 above)
Comment (sage-nathan-jones) [inline src/Hangfire.Core/Common/TypeExtensions.cs]: I don't think that I understand your suggestion, so I will present what my intents are.
Currently hangfire looks for a generic method signature by looking for the number of the methods generic types, but it populates these types by looking at the generic arguments. It cannot handle the case where there are types not in the argument list. (call this item "A") The loop at lines 79-90 are responsible for by calling TypesMatchRecursive which overwrites the values passed into the method's genericArguments variable. (item "B")

There is a bug in hangfire where if you try to use a generic method that has types which are not represented in the argument list of the method,(item "A" above) you get a NullReferenceException from the job invocation which is not helpful in resolving the matter. (item "C")

There seems to be no way to address this except for to include all of the generic types from the MethodInfo for serialization. (Hence the changes in InvocationData.) Once this metadata is included, then there is no need to "sniff" the generic types by looking at the argument list. (item "D")

So this PR tries to do multiple things: 

1) fix the issue by serializing all of the data needed and then using this info for invocation. Since this is taken from MethodInfo it should be considered correct over any other methods. 

2) It allows job serialized without the generic type metadata to still work within the constraints that are currently in place (namely arguments contain at least one of each generic type.) However, this sniffing will not be considered correct over explicit info (item "B" above). 

3) It changes the behavior of any pre-existing job that does not follow the constraint such that it does not throw a NullReferenceException in the job invocation, rather it behaves as if the method was not found (which is more accurate IMO). (item "C" above)

So I chose to leave the existing sniffing alone for maximum backwards compat. But only look at the sniffing in the case of missing explicit types. And even then I make sure that they're all there to prevent the unhelpful NullPointerException that we were originally experiencing.

With this context, could you rephrase your comment?

**Comment by sage-nathan-jones** [comment]
Ok, so I renamed the variables and moved around the logic as you suggested. I removed the code that was in place to check genericTypes against genericArguments since collapsing them to the same variable prohibits that (this was the .Zip() call.) This was a sanity test in any event. 

I also changed the .Aggregate() call to just be a foreach loop. This is still required in my opinion because without it, you will get the NullReferenceException during the MakeGenericMethod() call (which was what our original problem was.) This check has to be done after any sort of sniffing from TypeMatchRecursive().

**Comment by sage-nathan-jones** [comment]
I've made the requested changes. It has the original behavior of throwing a NullReferenceObject from MethodInfo.MakeGeneric()

**Comment by odinserj** [comment]
@sage-nathan-jones thanks a lot for your hard work! @aidmsu, to what branch should we merge these changes?

File: src/Hangfire.Core/Storage/InvocationData.cs [hunk 3]
```diff
@@ -30,17 +30,19 @@ namespace Hangfire.Storage
     public class InvocationData
     {
         public InvocationData(
-            string type, string method, string parameterTypes, string arguments)
```
Comment (aidmsu) [inline src/Hangfire.Core/Storage/InvocationData.cs]: I found one more problem. If we just add optional parameter `string genericTypes = null` to `InvocationData` constructor Hangfire.Core can become binary incompatible with Hangfire storage libraries such as Hangfire.Pro.Redis.

Instead I suggest to overload existing constructor and mark it by `JsonConstructorAttribute`:
```csharp
[JsonConstructor]
public InvocationData(string type, string method, string parameterTypes, string arguments, string genericTypes)
{
    Type = type;
    Method = method;
    ParameterTypes = parameterTypes;
    Arguments = arguments;
    GenericTypes = genericTypes;
}
```

**Comment by aidmsu** [comment]
@odinserj @sage-nathan-jones I found one more problem in the PR. Since this is bugfix I think we should merge these changes into master after fixing it.

File: src/Hangfire.Core/Storage/InvocationData.cs (hunk 1 above)
Comment (aidmsu) [inline src/Hangfire.Core/Storage/InvocationData.cs]: Ah, I missed an important thing(

If we add new `GenericTypes` property and create appropriate constructror `InvocationData(string type, string method, string parameterTypes, string arguments, string genericTypes)` we won't fix the bug for users who use [HangFire.Redis.StackExchange](https://github.com/marcoCasamento/Hangfire.Redis.StackExchange) or [Hangfire.Pro.Redis](https://nuget.hangfire.io/feeds/hangfire-pro/Hangfire.Pro.Redis/) for example. 

These packages don't use `JobHelper.ToJson` to serialize `InvocationData` to string and it won't store `GenericTypes` property automatically. Moreover in order to fix the bug for such users the new version of HangFire.Pro.Redis will have to depend on Hangfire.Core last version instead of Hangfire.Core 1.4.0. 

Therefore I suggest not to create separate property `GenericTypes` and store info about method generic parameters in `Method` property.

**Comment by odinserj** [comment]
@sage-nathan-jones, how are you doing? We realised that fix for this issue is a bit more complicated, because some storages (for example `Hangfire.Redis.StackExchange`) are using properties of the `InvocationData` class directly, without calls to `Serialize` or `Deserialize`. Therefore we can't add any properties here.

As a workaround, we can include the information regarding generic arguments to the `Method` property itself, for example `MyGenericMethod<System.String, System.Int32>`, but this will require additional parsing and much more effort for backward compatibility. 

I would like to know if you still want to complete this issue. If not – not a problem!

**Comment by sage-nathan-jones** [comment]
I will look at it this week. The simple cases should be straightforward, I worry about robustness... but since it is just flat out broken now, I guess it will be better than nothing.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
